doctr/datasets/svt.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (C) 2021, Mindee.
+
+# This program is licensed under the Apache License version 2.
+# See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
+
+import os
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Tuple
+from defusedxml import defuse_stdlib
+
+defuse_stdlib()
+import xml.etree.ElementTree as ET
+
+import numpy as np
+
+from .datasets import VisionDataset
+
+__all__ = ['SVT']
+
+
+class SVT(VisionDataset):
+    """SVT dataset from `"The Street View Text Dataset - UCSD Computer Vision"
+    <http://vision.ucsd.edu/~kai/svt/>`_.
+
+    Example::
+        >>> from doctr.datasets import SVT
+        >>> train_set = SVT(train=True, download=True)
+        >>> img, target = train_set[0]
+
+    Args:
+        train: whether the subset should be the training one
+        sample_transforms: composable transformations that will be applied to each image
+        rotated_bbox: whether polygons should be considered as rotated bounding box (instead of straight ones)
+        **kwargs: keyword arguments from `VisionDataset`.
+    """
+
+    URL = 'http://vision.ucsd.edu/~kai/svt/svt.zip'
+    SHA256 = '63b3d55e6b6d1e036e2a844a20c034fe3af3c32e4d914d6e0c4a3cd43df3bebf'
+
+    def __init__(
+        self,
+        train: bool = True,
+        sample_transforms: Optional[Callable[[Any], Any]] = None,
+        rotated_bbox: bool = False,
+        **kwargs: Any,
+    ) -> None:
+
+        super().__init__(self.URL, None, self.SHA256, True, **kwargs)
+        self.sample_transforms = sample_transforms
+        self.train = train
+        self.data: List[Tuple[Path, Dict[str, Any]]] = []
+        np_dtype = np.float16 if self.fp16 else np.float32
+
+        # Load xml data
+        tmp_root = os.path.join(self.root, 'svt1')
+        xml_tree = ET.parse(os.path.join(tmp_root, 'train.xml')) if self.train else ET.parse(
+            os.path.join(tmp_root, 'test.xml'))
+        xml_root = xml_tree.getroot()
+
+        for child in xml_root:
```

Comment:
1 child = 1 doc ? or 1 page ?

Reply:
1 child = 1 image path and all attributes but attributes are nested

Reply:
@fg-mindee

Reply:
if child = img, let's call the variable that way! But I'm confused,
those two lines:
```python
for child in xml_root:
```
and
```python
for image_tag in child:
```
seem to indicate that a child = list of images

Reply:
short example:
```
<tagset>
<image>
<imageName>img/18_01.jpg</imageName>
<address>? 216 Union Street Seattle WA</address>
<lex>THE,TRIPLE,DOOR,WILD,GINGER,ASIAN,RESTAURANT,GELATIAMO,SEATTLE,ART,MUSEUM,BENAROYA,HALL,POST,OFFICE,SYMPHONY,SHOWBOX,THE,MARKET,BYRNIE,UTZ,HATS,DOWNTOWN,EPICENTER,WASHINGTON,MUTUAL,INC,FINANCIAL,SERVICES,HARRIED,HUNGRY,CATERING,CAFE,ROSS,DRESS,FOR,LESS,REPUBLIC,PARKING,UNION,WILLIAM,TRAVER,GALLERY,MARCELLA,BOUTIQUE,SKINNER,BUILDING,BRUNO,MEXICAN,ITALIAN</lex>
<Resolution x="1280" y="800"/>
<taggedRectangles>
<taggedRectangle height="79" width="186" x="539" y="349">
<tag>DOOR</tag>
</taggedRectangle>
<taggedRectangle height="38" width="71" x="614" y="633">
<tag>THE</tag>
</taggedRectangle>
<taggedRectangle height="41" width="68" x="527" y="666">
<tag>THE</tag>
</taggedRectangle>
<taggedRectangle height="86" width="189" x="462" y="257">
<tag>TRIPLE</tag>
</taggedRectangle>
<taggedRectangle height="32" width="50" x="562" y="217">
<tag>THE</tag>
</taggedRectangle>
</taggedRectangles>
</image>
<image>
<imageName>img/13_11.jpg</imageName>
<address>1636 South 7th Street Phoenix AZ</address>
<lex>MAGIC,AUTO,SALES,ENTERPRISE,RENT,CAR,FIESTA,MOTORS,CAMACHO,COMPLETE,SERVICES,TOM,LOS,ARCOS,RESTAURANT,BAR,TACO,BELL,SAMYRA,LEBANESE,CUISINE,MOTEL,MANNY,DISCOUNT,REPAIR,KENZIE,ASSET,MANAGEMENT,1601,SOUTH,7TH,STREET,LLC,LLANTERA,FAMOSOS,BEST,WESTERN,MOTOR,LALO,CANTINA,AND,DANCING,THOMAS,PIPE,SUPPLY,PEREZ,TRIM,SHOP,COYOTE,KITCHEN,DINER,DIAMOND,PLATE,SERVICE,WEST,DELIVERY,EXPRESS</lex>
<Resolution x="1024" y="768"/>
<taggedRectangles>
<taggedRectangle height="34" width="106" x="474" y="452">
<tag>AUTO</tag>
</taggedRectangle>
<taggedRectangle height="31" width="120" x="351" y="455">
<tag>MAGIC</tag>
</taggedRectangle>
</taggedRectangles>
</image>
```

Reply:
root: tagset -> child from root: image -> tag from image: imageName

Reply:
alright but why are we iterating on image tags if there is only 1 per image ? :sweat_smile:

Reply:
no 1 image has nested attributes 
```
<Element 'imageName' at 0x7f64beda8130>
<Element 'address' at 0x7f64beda8180>
<Element 'lex' at 0x7f64beda8220>
<Element 'Resolution' at 0x7f64beda8270>
<Element 'taggedRectangles' at 0x7f64beda82c0>
```
and taggedRectangles have in one more depth a tag attribute     i know its confusing :smile:

Reply:
```
for image in xml_root:
    for image_attributes in image:
```
wdyt ?
@fg-mindee

Reply:
we can not call it in short attrib or else this would be bad as a result of that each XMLElement has tag, attrib and text

Reply:
And we can't access it by name?
like `child['imageName']`